ouroboros-consensus/src/Ouroboros/Consensus/Storage/ImmutableDB/API.hs

```diff
@@ -274,61 +231,273 @@ iteratorToList it = go []
         IteratorExhausted  -> return $ reverse acc
         IteratorResult res -> go (res:acc)
 
+-- | An iterator that is immediately exhausted.
+emptyIterator :: MonadSTM m => Iterator m blk b
+emptyIterator = Iterator {
+      iteratorNext    = return IteratorExhausted
+    , iteratorHasNext = return Nothing
+    , iteratorClose   = return ()
+    }
+
+{-------------------------------------------------------------------------------
+  Types
+-------------------------------------------------------------------------------}
+
+-- | Information about the tip of the ImmutableDB.
+data Tip blk = Tip {
+      tipSlotNo  :: !SlotNo
+    , tipIsEBB   :: !IsEBB
+    , tipBlockNo :: !BlockNo
+    , tipHash    :: !(HeaderHash blk)
+    }
+  deriving (Generic)
+
+deriving instance StandardHash blk => Eq                 (Tip blk)
+deriving instance StandardHash blk => Show               (Tip blk)
+deriving instance StandardHash blk => NoUnexpectedThunks (Tip blk)
+
+tipToRealPoint :: Tip blk -> RealPoint blk
+tipToRealPoint Tip { tipSlotNo, tipHash } = RealPoint tipSlotNo tipHash
+
+tipToPoint :: WithOrigin (Tip blk) -> Point blk
+tipToPoint = \case
+    Origin        -> GenesisPoint
+    NotOrigin tip -> realPointToPoint $ tipToRealPoint tip
+
+tipToAnchor :: WithOrigin (Tip blk) -> AF.Anchor blk
+tipToAnchor = \case
+    Origin ->
+      AF.AnchorGenesis
+    NotOrigin (Tip { tipSlotNo, tipHash, tipBlockNo }) ->
+      AF.Anchor tipSlotNo tipHash tipBlockNo
+
+blockToTip :: (HasHeader blk, GetHeader blk) => blk -> Tip blk
+blockToTip blk = Tip {
+      tipSlotNo  = blockSlot blk
+    , tipIsEBB   = blockToIsEBB blk
+    , tipBlockNo = blockNo blk
+    , tipHash    = blockHash blk
+    }
+
+-- | newtype with an 'Ord' instance that only uses 'tipSlotNo' and 'tipIsEBB'
+-- and ignores the other fields.
+newtype CompareTip blk = CompareTip { getCompareTip :: Tip blk }
+
+instance Eq (CompareTip blk) where
+  a == b = compare a b == EQ
+
+instance Ord (CompareTip blk) where
+  compare = mconcat [
+        compare      `on` tipSlotNo . getCompareTip
+      , compareIsEBB `on` tipIsEBB  . getCompareTip
+      ]
+    where
+      -- When a block and an EBB share a slot number, the EBB is "older".
+      compareIsEBB :: IsEBB -> IsEBB -> Ordering
+      compareIsEBB IsEBB    IsNotEBB = LT
+      compareIsEBB IsNotEBB IsEBB    = GT
+      compareIsEBB _        _        = EQ
+
+-- | Returned by 'streamBlocks' when a bound is wrong.
+--
+-- NOTE: this is not a 'UserError' that is thrown as an exception, but an
+-- admissible error, returned in an 'Either'. This is because we expect wrong
+-- bounds to be passed in practice. The functions to stream the blocks are
+-- best placed to validate the bounds, instead of requiring the users of the
+-- functions to do the validation beforehand.
+data WrongBoundError blk
```

Comment:
Rename it and use it as the argument to `MissingBlockError` (so `getKnownX` throws it too).